### PR TITLE
Native support for Websockets

### LINKED ISSUE #1491: Native support for web sockets
We've been attempting to migrate our app from sapper to sveltekit and ran into the problem of web sockets, in sapper we could simply attach the WebSocket server directly to the express server and be done with it, but as per Rich Harris' comments [here](https://github.com/sveltejs/kit/issues/334#issuecomment-804987028) and here [here](https://github.com/sveltejs/kit/issues/334#issuecomment-806289525) in sveltekit we don't want to expose direct access to the middlwares thus promoting the node adapter over the serverless ones.

However seeing as some of the serverless providers are starting to support web sockets ([cloudflare](https://blog.cloudflare.com/introducing-websockets-in-workers/)) ([begin](https://blog.begin.com/introducing-architect-5-0-fully-serverless-websockets-20bb97673c4e)) ([AWS Lambda](https://aws.amazon.com/blogs/compute/announcing-websocket-apis-in-amazon-api-gateway/)) should sveltekit provide a way to handle these connections directly? 

A native approach to this could also potentially provide a tool to solve this [issue](https://github.com/sveltejs/kit/issues/87) and [this](https://github.com/sveltejs/kit/issues/887)

**Describe the solution you'd like**
An interesting solution could be to expose functions in endpoints in the same way we handle http requests, and then let the adapters handle the rest

```routes/api/index.ts```
```js
// Handle post requests
export const post: RequestHandler = async (request) => {....}
// Handle web socket requests
export const ws: WebSocketRequestHandler = async (request) => {....}
```

A concern with this approach could be mixing protocols, within an endpoint could be confusing. Another could be if this becomes too complicated since as far as I know the initial [handshake](https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_servers#the_websocket_handshake) to start a WebSocket connection is handled over http, would this then be handled directly by sveltekit which could be considered too much "magic" since it would have to intercept a get request that would then not make it to the get handler in the endpoint. Or should we let the user implement the get endpoint and return the WebSocket pair and upgrade header.

**Describe alternatives you've considered**
The way we go about it today is a small script that manipulates the build output of sveltekit as such however this does not work in dev mode which is a pain point.
```js
const instance = createServer({ render: app.render }).listen(PORT, HOST, (err) => {
	if (err) {
		console.log('error', err);
	} else {
		console.log(`Listening on port ${PORT}`);
	}
});
app.server.installSubscriptionHandlers(instance);
export { instance };
```

Is this a feature that you feel would suit sveltekit? are there any considerations I've overlooked?

### LINKED ISSUE #12358: implement "Upgrade" as a function in a api route to support websocket
### Describe the problem

One can easy implement GET, PUT in an api route by a function (see below)
```javascript
import { json } from '@sveltejs/kit';

export function GET() {
	const number = Math.floor(Math.random() * 6) + 1;

	return json(number);
}
```

```

### Describe the proposed solution

If we implement UPGRADE we could implement websockets aswell

this file is in `src/route/chat-app/server.js`  connect with `new WebSocket('/chap-app')`

```javascript
// if this file is in src/route/chat-app/server.js

import { json } from '@sveltejs/kit';

// called when socket wants to upgrade
export function UPGRADE() {

      // this function will be called after successfull upgrade
       return function connection(ws: WebSocket) {
                         
 	       ws.send('hello');
               ws.onmessage = (event) => {..});
              // other handlers
      }
}

### Alternatives considered

_No response_

### Importance

would make my life easier

### Additional Information

Currently, customers expect some kind of live real time interaction,  a clean solution integrating with sveltekit would be a huge advantage

### PR DESCRIPTION
**Pull request by LukeHagar**
This PR is intended to add a path forward to natively support Websockets and similar protocols directly in SvelteKit by implementing an `UPGRADE` handler to be used similarly to existing `GET`, `POST`, `HEAD` `+server.js` exports

```ts
import { WebSocketServer } from 'ws';

const wss = new WebSocketServer({ noServer: true });

wss.on('connection', (ws) => {
	ws.on('error', console.error);

	ws.on('message', (message) => {
		console.log('received: %s', message);
	});
});

export function UPGRADE({ upgrade }) {
	wss.handleUpgrade(upgrade.request, upgrade.socket, upgrade.head, (ws) => {
		console.log('UPGRADED');
		wss.emit('connection', ws, upgrade.request);
	});
}
```

This PR will need to evolve to reflect the ideal implementation, and to correct things I have not yet like types, supports, etc...

<details open>
<summary>Resolves</summary>
Resolves #12358 

Resolves #1491 
</details>

---

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [ ] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [ ] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### Edits

- [ ] Please ensure that 'Allow edits from maintainers' is checked. PRs without this option may be closed.

### CODE REVIEW

File: packages/kit/src/runtime/server/endpoint.js [hunk 1]
```diff
@@ -40,28 +40,36 @@ export async function render_endpoint(event, mod, state) {
...
-		let response = await handler(
-			/** @type {import('@sveltejs/kit').RequestEvent<Record<string, any>>} */ (event)
-		);
-
-		if (!(response instanceof Response)) {
-			throw new Error(
-				`Invalid response from route ${event.url.pathname}: handler should return a Response object`
+		if (method === 'GET' && event.request.headers.has('upgrade') && event.upgrade && mod.UPGRADE) {
+			console.log('upgrade');
+			await mod.UPGRADE(/** @type {import('@sveltejs/kit').RequestEvent<Record<string, any>>} */ (event));
```
Comment (benmccann) [inline packages/kit/src/runtime/server/endpoint.js:45]: if you put a `return` after this line you could avoid wrapping everything that follows in an `else`

File: packages/kit/src/exports/vite/dev/index.js [hunk 2]
```diff
@@ -521,24 +567,28 @@ export async function dev(vite, vite_config, svelte_config) {
...
+							async_local_storage.enterWith({ event, config, prerender });
+						},
+						emulator
 					},
-					before_handle: (event, config, prerender) => {
-						async_local_storage.enterWith({ event, config, prerender });
-					},
-					emulator
-				});
+					undefined
```
Comment (benmccann) [inline packages/kit/src/exports/vite/dev/index.js:590]: I think you could remove this as leaving the last arg out would be equivalent
```suggestion
```

**Comment by Rich-Harris** [comment]
preview: https://svelte-dev-git-preview-kit-12961-svelte.vercel.app/

_this is an automated message_

File: packages/kit/src/runtime/server/endpoint.js (hunk 1 above)
Comment (LukeHagar) [reply packages/kit/src/runtime/server/endpoint.js]: Should I just simplify it to a conditional return?

File: packages/kit/src/runtime/server/respond.js [hunk 3]
```diff
@@ -54,9 +54,10 @@ const allowed_page_methods = new Set(['GET', 'HEAD', 'OPTIONS']);
  * @param {import('types').SSROptions} options
  * @param {import('@sveltejs/kit').SSRManifest} manifest
  * @param {import('types').SSRState} state
- * @returns {Promise<Response>}
+ * @param {{request: import('http').IncomingMessage, socket: import('stream').Duplex , head: Buffer}?} upgradeRequest
+ * @returns {Promise<void | Response>}
```
Comment (benmccann) [inline packages/kit/src/runtime/server/respond.js:58]: I'm not that familiar with `UPGRADE`, but is it true that it sends no response? A quick peek at MDN makes it appear to me that it does send a response

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Upgrade#overview

Comment (LukeHagar) [reply packages/kit/src/runtime/server/respond.js:58]: https://github.com/websockets/ws/blob/68614728c164eb55462ff12ba24cb30451d28ec6/lib/websocket-server.js#L379-L414

The response appears to be sent back through the socket itself.

Comment (LukeHagar) [reply packages/kit/src/runtime/server/respond.js:58]: There does appear to be some variance in the way the response is sent though, for example [Bun](https://bun.sh/docs/api/websockets#start-a-websocket-server) does not send a response, but [Deno](https://github.com/denoland/deno/blob/ef7432c03f83ad9e9ca2812d0ab5653e87fa5259/ext/http/02_websocket.ts#L40-L107) does appear to return a standard HTTP response.

Socket.IO also [returns a response](https://github.com/socketio/socket.io/blob/91e1c8b3584054db6072046404a24e79a17c1367/packages/engine.io/lib/server.ts#L782-L822) but they refuse to just give you the resp object, and instead insist on binding directly to a Server so they can send it themselves.

If we look at the [Node.JS server upgrade event](https://nodejs.org/api/http.html#event-upgrade), they specify the data is sent back through the socket, and their on upgrade handler does not even support sending a standard response object

Comment (LukeHagar) [reply packages/kit/src/runtime/server/respond.js:58]: I would love maintainer/community input here, its definitely a fun issue to try and solve for everyone :P

Comment (LukeHagar) [reply packages/kit/src/runtime/server/respond.js:58]: It looks like Socket.IO may actually support the same interface
https://github.com/socketio/socket.io/blob/91e1c8b3584054db6072046404a24e79a17c1367/packages/engine.io/README.md?plain=1#L40-L58

Comment (LukeHagar) [reply packages/kit/src/runtime/server/respond.js:58]: I have't had any luck getting this to work locally

**Comment by Rich-Harris** [comment]
Thank you, I'm excited for this functionality. I'm not quite sure this is the right API though. As @benmccann mentioned elsewhere, treating `UPGRADE` as an HTTP verb would be a little confusing.

I don't think it's just a naming issue though — I think this API is probably too low-level if I have to interact with `ws` myself, and too Node-centric. How would this work in the context of, say, [Cloudflare Workers](https://developers.cloudflare.com/workers/examples/websockets/) or [Durable Objects](https://developers.cloudflare.com/durable-objects/examples/websocket-server/) or [Bun](https://bun.sh/docs/api/websockets) or [Deno](https://docs.deno.com/examples/http-server-websocket/)?  It might be good to start with some concrete use cases and design an API around that; it may not even make sense to expose the socket directly at all given that it takes a subtly different form on each platform. For context, here's how [Nitro handles websockets](https://nitro.build/guide/websocket).

**Comment by LukeHagar** [comment]
Hey @Rich-Harris,

We are also discussing in the Svelte Discord [here](https://discord.com/channels/457912077277855764/750401468569354431/1303766381774241883). 

What do you think about exposing the [`hooks` detailed here](https://crossws.unjs.io/guide/hooks) as a new export from `hooks.server.js`? This approach would allow us to easily add a new entry to the manifest/options alongside the handle and hook up each adapter as needed.

**Comment by LukeHagar** [comment]
[Here is an example](https://github.com/sveltejs/kit/compare/main...LukeHagar:kit:crossws) of the above discussed solution.

This solution has some very specific differences which should be considered:
1. The lack of unique ws handlers per request, which may have been a necessary tradeoff for broader support anyway
2. The addition of an `upgrade` function similar to the current `read` function, this is intended to pass along the crossws instance along with any 'original' required request values that are provided by the serve function for different adapters (Bun, Deno, etc...)

That being said I don't think either of those things are bad, and I think that if we make the names of some variables or exports here a little more appealing (and get my bad types in order) this has the potential of being a very good solution.

**Comment by LukeHagar** [comment]
I take it back you can do different paths using [this feature](https://crossws.unjs.io/guide/resolver) of crossws

**Comment by loky-lp** [comment]
What about exposing exposing an [action-like](https://svelte.dev/docs/kit/form-actions#Default-actions) object in `+page.server.js/ts` files with the API similar to how [nitro does it](https://nitro.build/guide/websocket#usage)?

Something like:
```ts
import type { Socket } from './$types';

export const socket = {
	open: (peer) => {
		peer.send({ user: "server,", message: `Welcome ${peer}!` });
		peer.publish("chat", { user: "server", message: `${peer} joined!` });
		peer.subscribe("chat");
	},
	message: (peer, message) => {
		if (message.text().includes("ping")) {
			peer.send({ user: "server", message: "pong" });
		} else {
			const msg = {
				user: peer.toString(),
				message: message.toString(),
			};
			peer.send(msg); // echo
			peer.publish("chat", msg);
		}
	},
	close: (peer) => {
		peer.publish("chat", { user: "server", message: `${peer} left!` });
	},
} satisfies Socket;
```

This way WebSockets can be created on a per-route basis and each step (open, close and message) could be properly handled by each platform with all of its shenanigans (👀 looking at you AWS lamba)

Or even better, maybe this object could be passed directly to [crossws `defineHook`](https://crossws.unjs.io/guide/hooks)

**Comment by LukeHagar** [comment]
Seems like implementing things this way instead would still be pretty straightforward, we would just need to change the root usage to be the shape crossws uses for routing, and do some work in the middle to actually load the correct paths websocket handlers.

**Comment by benmccann** [comment]
> What about exposing exposing an [action-like](https://svelte.dev/docs/kit/form-actions#Default-actions) object in +page.server.js/ts files with the API similar to how [nitro does it](https://nitro.build/guide/websocket#usage)?

Yeah, I think that's a nice idea.

Looks like Nitro imports a per-adapter implementation of crossws, which is something we could do too. E.g.:
https://github.com/nitrojs/nitro/blob/4307202f5aa5099e53b1653f382df393e929e086/src/presets/deno/runtime/deno-server.ts#L8
https://github.com/nitrojs/nitro/blob/4307202f5aa5099e53b1653f382df393e929e086/src/presets/bun/runtime/bun.ts#L5

I also noticed that Nitro marked their feature as experimental: https://nitro.build/guide/websocket#opt-in-to-the-experimental-feature, which is something we could consider as well if we want a bit of flexibility in changing the API.

**Comment by LukeHagar** [comment]
Yes the current approach I am working on with crossws would I believe have the adapter import occurring at the SK adapter level. Allowing for the same approach.

**Comment by LukeHagar** [comment]
Closing in favor of new PR #12973

**Pull request closed without merging**